### PR TITLE
[SIL][PackageCMO] Fix dispatch thunk linker error and update table serialization  

### PR DESCRIPTION
**Pull request by elsh**
* Do not use [serialized_for_package] for witness method before Package CMO.
* Update v-table and witness-table and their entry serialization; allow non-serialized but visible entries in a serialized table so they can be optimized by other SIL opt passes.
* Only serialize MethodInst and KeyPathInst if it has the right visibility.

Resolves rdar://129089105&129088935

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci test

**Review by aschwaighofer — APPROVED** [review]
This looks reasonable to me.

acknowleding that there will be a future patch that:

- does not set the serialized attribute for package serialization in SILGen
- implement a systematic way (refactoring a type and a decl visitor for instructions as @slavapestov has eluded to earlier) to find instructions that block package serialization

this can be done in later PRs

File: lib/SILOptimizer/IPO/CrossModuleOptimization.cpp [hunk 1]
```diff
@@ -266,31 +264,81 @@ void CrossModuleOptimization::serializeFunctionsInModule(
...
-      // Then serialize the witness table itself.
-      wt.setSerializedKind(getRightSerializedKind(M));
+      // This checks if a wtable entry is not serialized and attempts to
+      // serialize (and its references) if they have the right visibility.
+      // This should not be necessary but is added to ensure all applicable
+      // symbols are serialized. Whether serialized or not is cached so
+      // this check shouldn't be expensive.
+      auto unserializedWTMethodRange = llvm::make_filter_range(wt.getEntries(), [&](auto &entry) {
+        return entry.getKind() == SILWitnessTable::Method &&
+        entry.getMethodWitness().Witness->getSerializedKind() != getRightSerializedKind(M);
```
Comment (nkcsgexi) [inline lib/SILOptimizer/IPO/CrossModuleOptimization.cpp:323]: nitpick: indentation

File: lib/SILOptimizer/IPO/CrossModuleOptimization.cpp [hunk 2]
```diff
@@ -266,31 +264,81 @@ void CrossModuleOptimization::serializeFunctionsInModule(
...
+
+      bool containsInternal = llvm::any_of(wtMethodsToSerialize, [&](auto &method) {
+        // If the entry is internal, wtable should not be serialized.
+        // However, if the entry is not serialized but has the right
+        // visibility, it can still be referenced, thus the vtable
+        // should serialized.
+        return !method->hasValidLinkageForFragileRef(getRightSerializedKind(M));
+      });
+      if (!containsInternal)
+        wt.setSerializedKind(getRightSerializedKind(M));
```
Comment (nkcsgexi) [inline lib/SILOptimizer/IPO/CrossModuleOptimization.cpp:341]: the logics for serializing v-table and w-table are very much alike. Can we extract a common function to handle both?

Comment (elsh) [reply lib/SILOptimizer/IPO/CrossModuleOptimization.cpp:356]: Some of these checks might not be necessary but added as a caution as mentioned in the comments; will follow up and remove them if found to be unnecessary.

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
> Does this PR #73902 need to be merged first? It seems this PR would have a dependency on it.

yup, merged.

**Comment by elsh** [comment]
@swift-ci test

**Pull request merged**